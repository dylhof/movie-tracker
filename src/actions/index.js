export const storeMovies = (movies) => ({
  type: 'STORE_MOVIES',
  movies
})

export const setCurrentUser = (name, id) => ({
  type: 'SET_USER',
  name,
  id
})

export const logoutCurrentUser = () => ({
  type: 'LOGOUT_USER'
})

export const addFavorite = (id) => ({
  type: 'ADD_FAVORITE',
  id
})

export const addAllUserFavorites = (favorites) => ({
  type: 'ADD_ALL_USER_FAVORITES',
  favorites
})

export const deleteFavorite = (id) => ({
  type: 'DELETE_FAVORITE',
  id
})

export const clearFavorites = () => ({
  type: 'CLEAR_FAVORITES',
})

export const setLoading = (bool) => ({
  type: 'SET_LOADING',
  bool
})