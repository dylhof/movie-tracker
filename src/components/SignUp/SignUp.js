import React, { Component } from 'react';

export default class SignUp extends Component {
  constructor() {
    super()
    this.state = {
      name: '',
      username: '',
      password: '',
      userFound: ''
    }
  }
  handleChange = (event) => {
    if (event.target.name === 'username') {
      this.setState({ username: event.target.value })
    } else if (event.target.password) {
      this.setState({ password: event.target.value })
    } else {
      this.setState({ name: event.target.value })
    }
  }

  //   sign in   /api/users
  // Create Account - /api/users / new
  //   Add Favorite - /api/users / favorites / new
  //   Receive All Favorites - /api/users /: user_id / favorites
  // Delete a Favorite - /api/users /: user_id / favorites /: movie_id

  handleSubmit = async (event) => {
    event.preventDefault();
    let allUsers
    try {
      const response = await fetch('http://localhost:3000/api/users')
      if (response.ok) {
        const result = await response.json()
        allUsers = result.data
      } else {
        throw Error(`There was an error: ${Error.status}`)
      }
    } catch (error) {

    }

    let foundUser = allUsers.find(user => {
      return user.email === this.state.username
    })

    if (foundUser) {
      this.setState({ userFound: 'User exists' })
    } else {



      const url = 'http://localhost:3000/api/users/new'
      try {
        const response = await fetch(url, {
          method: 'POST',
          body: JSON.stringify({ name: this.state.name, email: this.state.username, password: this.state.password }),
          headers: {
            'Content-Type': 'application/json'
          }
        })

        const result = await response.json()
        debugger
      }
      catch (error) {
        throw Error(`There was an error: ${Error.status}`)
      }
    }


  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <input name='name' value={this.state.name} onChange={this.handleChange} />
        <input name='username' value={this.state.username} onChange={this.handleChange} />
        <input name='password' value={this.state.password} onChange={this.handleChange} />
        <button>Submit</button>
        <span>{this.state.userFound}</span>
      </form>
    )
  }


}

// SignUp form:
// local state to store what they are typing in (control form)
// username field
// password field
// submit/ login button
// link to sign up form


// sign up form:
// local storage (control form)
// name field
// username field
// pw field
// submit/signup
// link to log in