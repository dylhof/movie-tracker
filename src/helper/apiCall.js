export const fetchData = async (url) => {
  const response = await fetch(url)
  if(response.ok) {
    return response.json() 
  } else {
    throw new Error(`Error fetching, code: ${response.status}`)
  }
}

export const fetchPost = async (url, options) => {
  const response = await fetch(url, options)
  if (response.ok) {
    return response.json()
  } else {
    throw new Error(response.status)
  }
}