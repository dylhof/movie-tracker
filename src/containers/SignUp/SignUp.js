import React, { Component } from 'react';
import { fetchPost } from '../../helper/apiCall';
import { connect } from 'react-redux';
import { setCurrentUser } from '../../actions';

export class SignUp extends Component {
  constructor() {
    super()
    this.state = {
      name: '',
      username: '',
      password: '',
      error: ''
    }
  }
  handleChange = (event) => {
    if (event.target.name === 'username') {
      this.setState({ username: event.target.value })
    } else if (event.target.name === 'password') {
      this.setState({ password: event.target.value })
    } else {
      this.setState({ name: event.target.value })
    }
  }

  //   sign in   /api/users
  // Create Account - /api/users / new
  //   Add Favorite - /api/users / favorites / new
  //   Receive All Favorites - /api/users /: user_id / favorites
  // Delete a Favorite - /api/users /: user_id / favorites /: movie_id

  handleSubmit = async (event) => {
    event.preventDefault();
    let allUsers
    try {
      const response = await fetchPost('http://localhost:3000/api/users/new',
        {
          method: 'POST',
          body: JSON.stringify({ name: this.state.name, email: this.state.username, password: this.state.password }),
          headers: {
            'Content-Type': 'application/json'
          }
        })
      this.props.dispatchSetCurrentUser(this.state.name, response.id)
      this.setState({name: '', username: '', password: ''})
    } catch (error) {
      console.log(error.message)
      if (error.message === '500') {
        this.setState({ error: 'This email already exists!' })
      }
    }
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <input name='name' value={this.state.name} onChange={this.handleChange} />
        <input name='username' value={this.state.username} onChange={this.handleChange} />
        <input name='password' value={this.state.password} onChange={this.handleChange} />
        <button>Submit</button>
        <span>{this.state.error}</span>
      </form>
    )
  }


}

export const mapDispatchToProps = (dispatch) => ({
  dispatchSetCurrentUser: (name, id) => dispatch(setCurrentUser(name, id))
})

export default connect(null, mapDispatchToProps)(SignUp)
// SignUp form:
// local state to store what they are typing in (control form)
// username field
// password field
// submit/ login button
// link to sign up form


// sign up form:
// local storage (control form)
// name field
// username field
// pw field
// submit/signup
// link to log in