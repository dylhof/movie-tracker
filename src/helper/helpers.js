import * as API from './apiCall';

export const tryUnfavorite = async (id, userID) => {

  const url = `http://localhost:3000/api/users/${userID}/favorites/${id}`
  await API.fetchPost(url,
    {
      method: 'DELETE',
      body: JSON.stringify({
        user_id: userID, movie_id: id
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

export const tryFavorite = async (title, id, userID, poster_path, release_date, vote_average, overview) => {

  await API.fetchPost('http://localhost:3000/api/users/favorites/new',
    {
      method: 'POST',
      body: JSON.stringify({
        movie_id: id, user_id: userID, title, poster_path,
        release_date, vote_average, overview
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

