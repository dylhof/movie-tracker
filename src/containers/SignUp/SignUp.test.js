import React from 'react'
import { shallow } from 'enzyme'
import { SignUp, mapDispatchToProps } from './SignUp'
import * as API from '../../helper/apiCall'
import * as actions from '../../actions/index'

describe('SignUp', () => {
  let wrapper
  beforeEach(() => {
    const mockDispatchSetCurrentUser = jest.fn()
    wrapper = shallow(
      <SignUp
        dispatchSetCurrentUser={mockDispatchSetCurrentUser}
      />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have initial State', () => {
    //setup
    const expected = { name: '', username: '', password: '' }
    //expectation
    expect(wrapper.state()).toEqual(expected)
  })
  
  describe('handleChange', () => {
    it('should set state of username', () => {
      //setup
      const mockEvent = { target: { name: 'username', value: 'Dylan' } }
      //execution
      wrapper.instance().handleChange(mockEvent)
      //expectation
      expect(wrapper.state('username')).toEqual('Dylan')
    })

    it('should set state of password', () => {
      //setup
      const mockEvent = { target: { name: 'password', value: 'Dylan' } }
      //execution
      wrapper.instance().handleChange(mockEvent)
      //expectation
      expect(wrapper.state('password')).toEqual('Dylan')
    })

    it('should set state of name', () => {
      //setup
      const mockEvent = { target: {name: 'name', value: 'Dylan'}}
      //execution
      wrapper.instance().handleChange(mockEvent)
      //expectation
      expect(wrapper.state('name')).toEqual('Dylan')
    })
  })

  describe('handleSubmit', () => {
    let mockEvent
    beforeEach(() => {
      mockEvent = Object.assign(jest.fn(), { preventDefault: () => { } })
    })
    it('should call fetchPost', () => {
      //setup
      API.fetchPost = jest.fn()
      //execution
      wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(API.fetchPost).toHaveBeenCalled()
    })

    it('should call dispatchSetCurrentUser with the correct params', async () => {
      //setup
      const mockResponse = { id: 2 }
      wrapper.instance().setState({ name: 'Matt' })
      API.fetchPost = jest.fn().mockImplementation(() => Promise.resolve(mockResponse))
      //execution
      await wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchSetCurrentUser).toHaveBeenCalledWith('Matt', 2)
    })

    it('should reset all state inputs to empty string', async () => {
      //setup
      wrapper.instance().setState({ name: 'Matt', username: 'Matt@matt', password: 'matt' })
      const expected = {name: '', username: '', password: ''}
      //execution
      await wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(wrapper.state()).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {

    it('should call dispatch when calling dispatchSetCurrentUser from mdtp', () => {
      //setup
      const mockDispatch = jest.fn()
      const actionToDispatch = actions.setCurrentUser('Matt', 1)
      //execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.dispatchSetCurrentUser('Matt', 1)
      //expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})