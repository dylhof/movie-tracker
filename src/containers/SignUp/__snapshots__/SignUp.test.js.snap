// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp should match the snapshot 1`] = `
<div
  className="signup-form-div"
>
  <form
    className="signup-form"
    onSubmit={[Function]}
  >
    <div
      className="signup-inner-div"
    />
    <div
      className="signup-inner-div"
    >
      <label
        className="signup-label"
        htmlFor="signup-name"
      >
        Name
      </label>
      <input
        className="signup-input"
        id="signup-name"
        name="name"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="signup-inner-div"
    >
      <label
        className="signup-label"
        htmlFor="signup-email"
      >
        Email
      </label>
      <input
        className="signup-input"
        id="signup-email"
        name="username"
        onChange={[Function]}
        value=""
      />
    </div>
    <div
      className="signup-inner-div"
    >
      <label
        className="signup-label"
        htmlFor="signup-password"
      >
        Password
      </label>
      <input
        className="signup-input"
        id="signup-password"
        name="password"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      className="signup-submit"
    >
      Submit
    </button>
  </form>
</div>
`;
